P2-978 - Feature flags in javascript

hansjovis opened the pull request:
## Context
<!--
What do we want to achieve with this PR? Why did we write this code?
-->

* We want a unified way to make feature flags available to the JavaScript side of the plugin (and any add-ons).

## Summary

<!--
Attach one of the following labels to the PR: `changelog: bugfix`, `changelog: enhancement`, `changelog: other`, `changelog: non-user-facing`.
If the changelog item is a bugfix, please use the following sentence structure: Fixes a bug where ... would ... (when ...).
If the changelog item is meant for the changelog of another add-on, start your changelog item with the name of that add-on's repo between square brackets, for example: * [wordpress-seo-premium] Fixes a bug where ....
If the changelog items is meant for the changelog of a javascript package, specify between square brackets in which package changelog the item should be included, for example: * [@yoast/components] Fixes a bug where ....
If the same changelog item is applicable to multiple changelogs/add-ons, add a separate changelog item for all of them.
-->
This PR can be summarized in the following changelog entry:

* Adds a unified way to make feature flags available to the JavaScript side of the plugin (and any add-ons).

## Relevant technical choices:

* I could not find a way to load the feature flags in the new way for the Elementor editor, so I decided to keep using the old way for now for the Elementor editor.
* **Note**: This will merge the `job-posting-step-2` feature branch into `trunk`. However, now it will be behind the `YOAST_SEO_SCHEMA_BLOCKS` feature flag.
* Added `--no-scripts` to Travis config when installing 7.0 dependencies, since we do not want to compile the DI-container (it would fail on variadic constructors). `--no-scripts` prevents the execution of all scripts defined in `composer.json`.

## Test instructions
<!--
Please follow these guidelines when creating test instructions:
- Please provide step-by-step instructions how to reproduce the issue, if applicable.
- Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->
### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

* Make sure that one or more feature flags have been enabled on your WordPress installation.
	* e.g. by adding `define( 'YOAST_SEO_SCHEMA_BLOCKS', true );` to your installation's `wordpress-config.php` file.

#### Making sure that the feature flags are available in the browser.
* Edit a post.
* In the browser console, enter this command:
  ```js
  window.yoast.featureFlag.enabledFeatures()
  ```
  * This should give you a list of enabled feature flags. If you enabled just the `YOAST_SEO_SCHEMA_BLOCKS` flag, this should return `[ 'SCHEMA_BLOCKS' ]`.

#### Making sure that the feature flags are available in the analysis web worker.
* Edit a post.
* In the browser console, switch to the analysis worker JavaScript context.
	* You can do this by clicking on the dropdown at the top of the console (that currently says 'top') and selecting the item that starts with 'analysis-worker'.
* In the browser console, enter this command:
  ```js
  self.yoast.featureFlag.enabledFeatures()
  ```
  * This should give you a list of enabled feature flags. If you enabled just the `YOAST_SEO_SCHEMA_BLOCKS` flag, this should return `[ 'SCHEMA_BLOCKS' ]`.

#### Making sure that the feature flags are available on term pages.
* Follow the steps above (for in the browser and in the analysis web worker) but now on term pages.
	* E.g. create a new category and edit its contents. 

### Test instructions for QA when the code is in the RC
<!--
Sometimes some steps from the test instructions for the acceptance test aren't relevant anymore once the code has been merged or the feature is complete. If that is the case, do not check the checkbox below.
QA is our Quality Assurance team. The RC is the release candidate zip that is tested before a release 
-->

* [x] QA should use the same steps as above.

<!--
If the above checkbox has not been checked, write down all steps QA should take to test this PR, not only the difference with the acceptance test steps. If QA should use the test instructions specified on the epic, paste a link to the relevant comment on the epic.
-->
QA can test this PR by following these steps:

*

## Impact check
<!--
Sometimes PRs have a bigger impact than is suggested in the user-facing changes. In such cases,
additional (regression) testing might be necessary. To make it clear what parts might need additional testing, please outline which parts of the plugin have been impacted by this PR.
-->
This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes

* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Documentation

* [ ] I have written documentation for this change.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [x] I have added unittests to verify the code works as intended

Fixes #

johannadevos commented:
CR: overall a very nice addition to the plugin(s)! I just have a few minor questions and remarks about test instructions.

- [ ] I'm missing test instructions to test whether this PR has the desired effect, i.e. set `YOAST_SEO_SCHEMA_BLOCKS` to `false` (or remove it entirely) and see that the job block templates aren't loaded. Also the effects on the SchemaTab in the metabox when enabling/disabling the feature flag. Make sure all your code changes are reflected in test instructions.
- [ ] I'm not 100% sure the changes in `class-metabox` and `class-taxonomy` are covered in the current test instructions. Could you please double check?

johannadevos commented:
Note: this PR is easier to review by temporarily basing it on `job-posting-step-2`.

hansjovis commented on inc/class-wpseo-utils.php:
```diff
@@ -1148,14 +1150,13 @@ public static function extend_kses_post_with_forms( $allowed_post_tags ) {
+		/**
+		 * The feature flag integration.
+		 *
+		 * @var Feature_Flag_Integration $feature_flag_integration;
```
You are right. However, without it PHPStorm does not recognize that the `$feature_flag_integration` variable is of type `Feature_Flag_Integration`.

hansjovis commented on packages/js/src/containers/SchemaTab.js:
```diff
@@ -95,19 +96,27 @@ export default compose( [
 			defaultPageType: getDefaultPageType(),
-			requiredBlockNames: getRequiredBlockNames(),
-			recommendedBlockNames: getRecommendedBlockNames(),
+			requiredBlockNames,
```
Yes and no. 

This syntax:
```js
const x = {
	param1: 'string',
	param2,
};
```
is the same as:
```js
const x = {
	param1: 'string',
	param2: param2,
};
```

I am thinking about using less of this 'syntactic sugar' in my code though. Although it is terse, it ultimately makes the code more difficult to understand.

johannadevos replied:
That's not a bad idea - I think readability should come before terseness.
(Also thanks for explaining the syntax!) 🙂

hansjovis commented on src/integrations/feature-flag-integration.php:
```diff
@@ -0,0 +1,102 @@
+		 */
+		$filtered_enabled_features = apply_filters( 'wpseo_enable_feature', $enabled_features );
+
+		if ( ! is_array( $filtered_enabled_features ) || empty( $filtered_enabled_features ) ) {
```
Good point.

There is no harm in returning an empty array here. I am going to remove the `empty` check.

hansjovis commented on tests/unit/integrations/feature-flag-integration-test.php:
```diff
@@ -0,0 +1,224 @@
+		$this->instance = new Feature_Flag_Integration( $this->asset_manager, ...$this->feature_flag_conditionals );
+	}
+
+	/**
```
I added a unit test for the constructor and a couple of other tests to cover some previously uncovered paths (mostly when the filter returns invalid output).

hansjovis commented:
@johannadevos I now added test instructions for the job block templates to the PR for WordPress SEO Premium (since they are Premium facing changes).

I also added to the test instructions on this PR that they should also be tested on category/term pages. That should cover the changes in `class-taxonomy`. The changes in `class-metabox` were already tested by the original test instructions.

Thanks for the thorough review!

johannadevos commented on .travis.yml:
```diff
@@ -153,8 +153,8 @@ install:
       # Run composer update as we have dev dependencies locked at PHP ^7.0 versions.
-      travis_retry composer update
-      travis_retry composer install --no-interaction
+      travis_retry composer update --no-interaction --no-scripts
```
Could you explain why you made these changes?

johannadevos replied:
Found the answer in the Premium PR: "Since we do not want to compile the DI-container (it would fail on variadic constructors)." Will add it to the technical choices.

johannadevos commented:
CR on the latest commits: ✅

johannadevos commented:
Acceptance: ✅ 

For posts:
- [x] Making sure that the feature flags are available in the browser.
- [x] Making sure that the feature flags are available in the analysis web worker.

For terms:
- [x] Making sure that the feature flags are available in the browser.
- [x] Making sure that the feature flags are available in the analysis web worker.
